core/lib/src/shutdown.rs

```diff
@@ -0,0 +1,78 @@
+use futures::{channel::mpsc, stream::StreamExt};
+use std::{
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+/// Wrapper around an mpsc channel.
+#[derive(Debug)]
+crate struct Shutdown {
```

Comment:
I think we can avoid this struct and impls: we can inline the `mpsc::channel(1)` construction into `rocket.rs` (storing the Option<Receiver> directly in `Rocket`) and then use `.with_graceful_shutdown(async move { receiver.next().await; })`. That should save quite a bit of code.

Reply:
Should just be a slight compilation time improvement, as I imagine rustc is smart enough to inline it. Can do, though.